Move analyzer before detector (#197)
buildpacks/spec: platform.md

      - [Rebase](#rebase)
      - [Launch](#launch)
    - [Usage](#usage)
      - [`analyzer`](#analyzer)
        - [Inputs](#inputs-1)
        - [Outputs](#outputs-1)

>         - [Layer analysis](#layer-analysis)

Reviewer: I assume we need to move this into the restorer section now?

Reviewer: Or just roll it all under `Layer Restoration`(currently `Layer restoration` but it really should be capitalized)

Author: I rolled it under Layer Restoration

      - [`detector`](#detector)
        - [Inputs](#inputs)

>         - [Outputs](#outputs)

Reviewer: You can use the VScode plugin mentioned in the [comment](https://github.com/buildpacks/spec/pull/197/files#diff-e603760990971da3f77be4bb8d77c3405098f006814fd8c054d2d15f395b8330R15) above the TOC to regenerate consistent anchor links (e.g. `#outputs-2`).

- MUST read `CNB_PLATFORM_API` from the execution environment and evaluate compatibility before attempting to parse other inputs (see [Platform API Compatibility](#platform-api-compatibility))
- MUST give command line inputs precedence over other inputs

#### `analyzer`
Usage:
```
/cnb/lifecycle/analyzer \
  [-analyzed <analyzed>] \
  [-daemon] \ # sets <daemon>
  [-gid <gid>] \

>   [-group <group>] \

Reviewer 2: From the lifecycle PR: it looks like group is only defined for earlier platform apis, is that intended?

Reviewer 2: Now that I'm remembering how all of this works... we can't have group if analyze goes first...
```suggestion
```

Author: We can have `group`, but not `order`

Reviewer 3: I think the future we will support `stack-group` in order to validate mixins, but not `group`?

Author: I think I confused myself with the names. We need `<order>`, not `<group`>. Fixed in 6480d2e975ba81d5b61d681f0177539408c834b8

- MUST read `CNB_PLATFORM_API` from the execution environment and evaluate compatibility before attempting to parse other inputs (see [Platform API Compatibility](#platform-api-compatibility))
- MUST give command line inputs precedence over other inputs

#### `analyzer`
Usage:
```
/cnb/lifecycle/analyzer \
  [-analyzed <analyzed>] \

>   [-cache-image <cache-image>] \

Reviewer: ```suggestion
```
We aren't using this flag yet, so let's remove it.

  [-daemon] \ # sets <daemon>
  [-gid <gid>] \
  [-group <group>] \
  [-layers <layers>] \
  [-log-level <log-level>] \
  [-previous-image <previous-image> ] \
  [-uid <uid>]
```

##### Inputs
| Input             | Environment Variable  | Default Value            | Description
|-------------------|-----------------------|--------------------------|----------------------
| `<analyzed>`      | `CNB_ANALYZED_PATH`   | `<layers>/analyzed.toml` | Path to output analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)

> | `<cache-image>`   | `CNB_CACHE_IMAGE`     |                          | Location of cache, provided as an image

Reviewer: ```suggestion
```

- MUST read `CNB_PLATFORM_API` from the execution environment and evaluate compatibility before attempting to parse other inputs (see [Platform API Compatibility](#platform-api-compatibility))
- MUST give command line inputs precedence over other inputs

#### `analyzer`
Usage:
```
/cnb/lifecycle/analyzer \
  [-analyzed <analyzed>] \
  [-daemon] \ # sets <daemon>
  [-gid <gid>] \
  [-group <group>] \
  [-layers <layers>] \
  [-log-level <log-level>] \
  [-previous-image <previous-image> ] \
  [-run-image <run-image> ] \

>   [-stack-id <stack-id> ] \

Author: Removed in 6480d2e975ba81d5b61d681f0177539408c834b8

  [-stack <stack> ] \
  [-tag <tag>...] \
  [-uid <uid>] \
  <image>
```

##### Inputs
| Input             | Environment Variable  | Default Value            | Description
|-------------------|-----------------------|--------------------------|----------------------
| `<analyzed>`      | `CNB_ANALYZED_PATH`   | `<layers>/analyzed.toml` | Path to output analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)
| `<daemon>`        | `CNB_USE_DAEMON`      | `false`                  | Analyze image from docker daemon
| `<gid>`           | `CNB_GROUP_ID`        |                          | Primary GID of the stack `User`

> | `<group>`         | `CNB_GROUP_PATH`      | `<layers>/group.toml`    | Path to group definition (see [`group.toml`](#grouptoml-toml))

Author: Removed in 6480d2e975ba81d5b61d681f0177539408c834b8

> | `<layers>`        | `CNB_LAYERS_DIR`      | `/layers`                | Path to layers directory

Reviewer: ```suggestion
| `<image>`           |                            |                     | Tag reference to which the app image will be written
| `<layers>`        | `CNB_LAYERS_DIR`      | `/layers`                | Path to layers directory
```

Author: Added in 6480d2e975ba81d5b61d681f0177539408c834b8

| `<log-level>`     | `CNB_LOG_LEVEL`       | `info`                   | Log Level
| `<previous-image>`| `CNB_PREVIOUS_IMAGE`  | `<image>`                | Image reference to be analyzed (usually the result of the previous build)
| `<run-image>`     | `CNB_RUN_IMAGE`       | resolved from <stack>    | Run image reference
| `<stack>`         | `CNB_STACK_PATH`      | `/cnb/stack.toml`        | Path to stack file (see [`stack.toml`](#stacktoml-toml))
| `<tag>...`        |                       |                          | Additional tag to apply to exported image
| `<uid>`           | `CNB_USER_ID`         |                          | UID of the stack `User`

- **If** `<daemon>` is `false`, `<image>` MUST be a valid image reference

> - **If** `<daemon>` is `true`, `<image>` MUST be either a valid image reference or an imageID

Author: Added in 6480d2e975ba81d5b61d681f0177539408c834b8

> - **If** `<run-image>` is not provided by the platform the value will be resolved from the contents of stack

Author: Added in 6480d2e975ba81d5b61d681f0177539408c834b8

- The lifecycle MUST accept valid references to non-existent images without error.
- The lifecycle MUST ensure that the `<run-image>` is compatible with the `<previous-image>`. To be compatible, the following conditions must be met:
  - Stack IDs must match
  - Platform/Architecture must be the same

>   - `<run-image>` `mixins` must be a superset of `<previous-image>` mixins

Author: Updated in 6480d2e975ba81d5b61d681f0177539408c834b8

> - The lifecycle MUST ensure registry read/write access to `<run-image>`, `<previous-image>`, `<tag>`

Author: Added in 6480d2e975ba81d5b61d681f0177539408c834b8

##### Outputs
| Output             | Description
|--------------------|----------------------------------------------
| [exit status]      | (see Exit Code table below for values)
| `/dev/stdout`      | Logs (info)
| `/dev/stderr`      | Logs (warnings, errors)

> | `<analyzed>`       | Analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)

Reviewer: Now that we are writing `run-image` metadata in this output. `analyzed.toml` needs to be an input to `exporter` and `stack.toml` can be removed as an exporter input.

Author: `<analyzed>` is already an input, I removed `<stack>` in 6480d2e975ba81d5b61d681f0177539408c834b8

| Exit Code | Result|
|-----------|-------|
| `0`       | Success
| `11`      | Platform API incompatibility error
| `12`      | Buildpack API incompatibility error
| `1-10`, `13-99` | Generic lifecycle errors

> | `200-299` | Analysis-specific lifecycle errors

Reviewer 4: ```suggestion
| `30-39` | Analysis-specific lifecycle errors
```

Update codes to match #196

Reviewer 4: Do we want to change the code numbers based off ordering?

Author: I don't think so - the order will be different depending on your platform API version selected for some amount of time.

> - The lifecycle MUST write [analysis metadata](#analyzedtoml-toml) to `<analyzed>`.

Reviewer: ```suggestion

- The lifecycle MUST write [analysis metadata](#analyzedtoml-toml) to `<analyzed>`, where:
  -  `build-image` MUST match the values provided in `<stack>`
  -  `run-image` MUST describe the `<run-image>`
  - **If** the `<previous-image>` is compatible with the `<run-image>` `previous-image` MUST describe `<previous-image>`.  `<previous-image>` is compatible with `<run-image>` if:
    - The stack IDs match
    - `<run-image>` `mixins` are a superset of `<previous-image>` mixins
  - **Else** the lifecycle MUST omit `previous-image.reference` and `previous-image.metadata`
```

Author: Added in 6480d2e975ba81d5b61d681f0177539408c834b8

- MUST read `CNB_PLATFORM_API` from the execution environment and evaluate compatibility before attempting to parse other inputs (see [Platform API Compatibility](#platform-api-compatibility))
- MUST give command line inputs precedence over other inputs

#### `analyzer`

> The platform MUST execute `analyzer` in the **build environment**

Reviewer: The way we define "build environment" I don't believe this is true. You could still run analyzer on the lifecycle image for example, it is not necessary that the analyze container extend from the build image.

Author: @ekcasey to validate mixins against all buildpack defined mixins in this phase, won't we need to be on the build image?

Author: I removed this and added a line to detector about validating mixins.

Usage:
```
/cnb/lifecycle/analyzer \
  [-analyzed <analyzed>] \
  [-daemon] \ # sets <daemon>
  [-gid <gid>] \
  [-layers <layers>] \
  [-log-level <log-level>] \

>   [-order <order>] \

Reviewer 2: What's the order used for?

Author: I think the intent is for validation purposes. So we can, for instance, fail a build if any of the buildpacks in the provided order request a mixin that the stackpacks can't provide. We could also move up the buildpack API validation to analyzer as well.

Reviewer 2: Oh hmmm... but wouldn't it make sense to do that check during detect? What if a buildpack requests a mixin that is not provided, but is going to fail detect anyway? I feel like it only makes sense to fail if the buildpack detects and is non-optional. Was this outlined in the RFC?

Reviewer 3: Yes, this is captured in the [What it is](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md#what-it-is) section of the RFC.

>  Before any lifecycle phases:
>
 >   The platform may compare the list of mixins that are statically required by all buildpacks (in the stacks sections of their buildpack.toml files) with the static list of mixins provided by the stack images and the stack buildpacks (in the stacks sections of their buildpack.toml files), and fail the build if the required mixins cannot be provided.

Author: This also matches the way `pack` validates IIRC. They currently validate mixins prior to executing anything, we we are sort of matching. The thought was that we should aggressively validate up front and relax later if there is desire to do so. The opposite is harder to sell since it would be a breaking change.

Reviewer 2: Makes sense, thanks!

  [-previous-image <previous-image> ] \

>   [-run-image <run-image> ] \

Reviewer 2: Do we need `-run-image` if we have `stack`?

Reviewer 2: Also why would we allow run-image to be overridden but not build-image?

Author: I think so, yes. Since a run image could exist on any registry, including private registries, I think we want to continue to support taking in a `run-image`. 

Phases after this should maybe be able to remove it though, right @ekcasey ?

I think we could remove `<run-image>` from `/builder` entirely since the `/analyzer` will produce an `analyzed.toml` with the image ref for the chosen `<run-image>`.

Reviewer 2: I guess I'm still confused why you couldn't pass in the build-image. What if `/cnb/stack.toml` doesn't contain it? Should the build fail?

Author: You can't pass in the build-image because you are running _on_ the build image for analyzer now.

Reviewer: @jabrown85 I don't think we should require that you run analyze on the build image, only that `stack.toml` accurately describes the build image.

| `<gid>`           | `CNB_GROUP_ID`        |                          | Primary GID of the stack `User`
| `<layers>`        | `CNB_LAYERS_DIR`      | `/layers`                | Path to layers directory
| `<image>`         |                       |                          | Tag reference to which the app image will be written
| `<log-level>`     | `CNB_LOG_LEVEL`       | `info`                   | Log Level
| `<order>`         | `CNB_ORDER_PATH`      | `/cnb/order.toml`        | Path to order definition (see [`order.toml`](#ordertoml-toml))
| `<previous-image>`| `CNB_PREVIOUS_IMAGE`  | `<image>`                | Image reference to be analyzed (usually the result of the previous build)
| `<run-image>`     | `CNB_RUN_IMAGE`       | resolved from <stack>    | Run image reference
| `<stack>`         | `CNB_STACK_PATH`      | `/cnb/stack.toml`        | Path to stack file (see [`stack.toml`](#stacktoml-toml))
| `<tag>...`        |                       |                          | Additional tag to apply to exported image
| `<uid>`           | `CNB_USER_ID`         |                          | UID of the stack `User`

> - **If** `<daemon>` is `false`, `<image>` MUST be a valid image reference
> - **If** `<daemon>` is `true`, `<image>` MUST be either a valid image reference or an imageID
> - **If** the platform provides one or more `<tag>` inputs, each `<tag>` MUST be a valid image reference.

Reviewer: ```suggestion
-`<image>` MUST be a valid image reference
- **If** the platform provides one or more `<tag>` inputs, each `<tag>` MUST be a valid image reference.
- **If** `<daemon>` is `false`, and `<previous-image>`, if provided,  MUST be a valid image reference
- **If** `<daemon>` is `true`, `<previous-image>`, if provided, MUST be either a valid image reference or an imageID
```
Sorry! I screwed this up the first time. `<image>` can't be an imageID in the daemon case, b/c we won't be writing to an imageID, this restriction should instead apply to the `<previous-image>`.

| `<layers>`        | `CNB_LAYERS_DIR`      | `/layers`                | Path to layers directory
| `<image>`         |                       |                          | Tag reference to which the app image will be written
| `<log-level>`     | `CNB_LOG_LEVEL`       | `info`                   | Log Level
| `<order>`         | `CNB_ORDER_PATH`      | `/cnb/order.toml`        | Path to order definition (see [`order.toml`](#ordertoml-toml))
| `<previous-image>`| `CNB_PREVIOUS_IMAGE`  | `<image>`                | Image reference to be analyzed (usually the result of the previous build)
| `<run-image>`     | `CNB_RUN_IMAGE`       | resolved from <stack>    | Run image reference
| `<stack>`         | `CNB_STACK_PATH`      | `/cnb/stack.toml`        | Path to stack file (see [`stack.toml`](#stacktoml-toml))
| `<tag>...`        |                       |                          | Additional tag to apply to exported image
| `<uid>`           | `CNB_USER_ID`         |                          | UID of the stack `User`

-`<image>` MUST be a valid image reference
- **If** the platform provides one or more `<tag>` inputs, each `<tag>` MUST be a valid image reference.

> - **If** `<daemon>` is `false`, and `<previous-image>`, if provided,  MUST be a valid image reference

Reviewer: ```suggestion
- **If** `<daemon>` is `false`, `<previous-image>`, if provided,  MUST be a valid image reference
```

- MUST read `CNB_PLATFORM_API` from the execution environment and evaluate compatibility before attempting to parse other inputs (see [Platform API Compatibility](#platform-api-compatibility))
- MUST give command line inputs precedence over other inputs

#### `analyzer`
Usage: 
```

> /cnb/lifecycle/analyzer \

Reviewer: The RFC mentions the addition of `-run-image`, destination tags, `stack`, and `cache-dir`. Should we be including these and the associated behavior (recording the run-image digest, validating write access etc.) in this PR? If not should we be creating separate issues for API 0.7?

Author: I wasn't _planning_ on adding those inputs in this update. I can create tickets. I could also be convinced we should do this now.

Author: I added the additional fields in https://github.com/buildpacks/spec/pull/197/commits/de9d5a694204abccfae24a746411e4939cf845c9.

Author: I re-removed these fields. Will followup with another spec PR to add them back and discuss what exactly we do with those fields.

  [-analyzed <analyzed>] \
  [-cache-image <cache-image>] \
  [-daemon] \ # sets <daemon>
  [-gid <gid>] \
  [-group <group>] \
  [-layers <layers>] \
  [-log-level <log-level>] \
  [-uid <uid>] \

>   <image>

Reviewer: After looking over the `creator` usage and thinking about a future where we might want to validate write access to the destination tags in `analyzer` I think we should replace the `<image>` positional argument with an optional `-previous-image` flag. If we later introduce a positional argument to designate the destination tag we can default `previous-image` to image like we do in `creator`.

We typically use the positional arguments for required configuration. The previous image was formerly required by analyzer, but I think that is unintuitive on a first build as we should take the opportunity to change it while we are changing the usage here.

Should we do this now?:
Pros: It will be easier for platform implementers if we bundle more changes together
Cons: It will delay the next lifecycle release

Author: I'm fine with making this change now. I'll update the spec

Author: Hmm - I guess this would look like `-previous-image` needing to be required for Platform API < 0.6?

Author: I got this working in the PR. Commit here https://github.com/buildpacks/lifecycle/pull/505/commits/5988b8a58c818d4d4a6032d78b931fcc42548cdc

It could probably use some 👀  and potential spec updates to reflect the new behavior. Like when you don't have a `-previous-image`, do we need to explain which parts will and will not occur? I didn't write any tests for that yet. For `0.6` _right now_ omitting `-previous-image` and would result in an empty `analyzed.toml` I think. No metadata or image id would exist in that case. Obviously we want to start storing things about the run-image including the stack/mixins so detector and other phases could consume that information analzyed in this first step.

Author: oh, just kidding. I'll need to make sure in 0.5 and below the positional argument is allowed/required like it is today and not allow `-previous-image`. I'll fix that up tomorrow hopefully

|--------------------|----------------------------------------------
| [exit status]      | (see Exit Code table below for values)
| `/dev/stdout`      | Logs (info)
| `/dev/stderr`      | Logs (warnings, errors)
| `<group>`          | Detected buildpack group  (see [`group.toml`](#grouptoml-toml))
| `<plan>`           | Resolved Build Plan (see [`plan.toml`](#plantoml-toml))

| Exit Code | Result|
|-----------|-------|
| `0`       | Success
| `11`      | Platform API incompatibility error
| `12`      | Buildpack API incompatibility error
| `1-10`, `13-99` | Generic lifecycle errors
| `100`     | All buildpacks groups have failed to detect w/o error
| `101`     | All buildpack groups have failed to detect and at least one buildpack has errored
| `102-199` | Detection-specific lifecycle errors

The lifecycle:
- SHALL detect a single group from `<order>` and write it to `<group>` using the [detection process](buildpack.md#phase-1-detection) outlined in the Buildpack Interface Specification
- SHALL write the resolved build plan from the detected group to `<plan>`

#### `restorer`
Usage:
```
/cnb/lifecycle/restorer \
  [-cache-dir <cache-dir>] \
  [-cache-image <cache-image>] \

>   [-daemon] \ # sets <daemon> 

Reviewer: I think this gets confusing in conjunction with our registry only cache image. Maybe we can use the analysis metadata to record whether the previous image is a daemon or registry image?

Also, this might be a really good opportunity to make a long desired change and move from have a single `reference` key in [analyzed.toml](https://github.com/buildpacks/spec/blob/main/platform.md#analyzedtoml-toml) that stores either an imageID or digest reference, to having two explicit keys (`image_id`, `registry_digest`)

Reviewer: Apologies I thought I updated this comment with an edit. I realized existing `analyzer` already has this issue (if you want to call it an issue) and this suggestion should probably be out of scope for now. But I wanted to leave these thoughts somewhere, in case we want to incorporate them in future changes.

  [-gid <gid>] \
  [-group <group>] \
  [-layers <layers>] \
  [-log-level <log-level>] \
  [-skip-layers <skip-layers>] \
  [-uid <uid>]
```

##### Inputs
| Input          | Environment Variable  | Default Value            | Description
|----------------|-----------------------|--------------------------|----------------------
| `<analyzed>`   | `CNB_ANALYZED_PATH`   | `<layers>/analyzed.toml` | Path to analysis metadata (see [`analyzed.toml`](#analyzedtoml-toml)
| `<cache-dir>`  | `CNB_CACHE_DIR`       |                          | Path to a cache directory
| `<cache-image>`| `CNB_CACHE_IMAGE`     |                          | Reference to a cache image in an OCI image registry
| `<gid>`        | `CNB_GROUP_ID`        |                          | Primary GID of the stack `User`
| `<group>`      | `CNB_GROUP_PATH`      | `<layers>/group.toml`    | Path to group definition (see [`group.toml`](#grouptoml-toml))
| `<layers>`     | `CNB_LAYERS_DIR`      | `/layers`                | Path to layers directory
| `<log-level>`  | `CNB_LOG_LEVEL`       | `info`                   | Log Level
| `<uid>`        | `CNB_USER_ID`         |                          | UID of the stack `User`

> | `<skip-layers>`| `CNB_SKIP_LAYERS`     | `false`                  | Do not perform layer analysis

Reviewer: ```suggestion
| `<skip-layers>`| `CNB_SKIP_LAYERS`     | `false`                  | Do not perform [layer restoration]((#layer-restoration)
```

##### Outputs
| Output                                | Description
|---------------------------------------|----------------------------------------------
| [exit status]                         | (see Exit Code table below for values)
| `/dev/stdout`                         | Logs (info)

> | `/dev/stderr`                         | Logs (warnings, errors)

Reviewer: ```suggestion
| `/dev/stderr`                         | Logs (warnings, errors)
| `<layers>/<buidpack-id>/store.toml`   | Persistent metadata (see data format in [Buildpack Interface Specification](buildpack.md))
```

Reviewer: This was missing from the spec before, but while we are looking at `analyzer` and `restorer`, we should add it.

> | `<layers>/<buidpack-id>/<layer>.sha`  | Files containing the diffID of each analyzed layer

Reviewer: ```suggestion
```
We don't need `<layer>.sha` files anymore. They previously existed to pass layer diffIDs between `analyzer` and `restorer` but they we always a little janky and are now, thankfully, unnecessary.

| `<layers>/<buidpack-id>/<layer>.toml` | Files containing the layer content metadata of each analyzed layer (see data format in [Buildpack Interface Specification](buildpack.md))
| `<layers>/<buidpack-id>/<layer>/*`.   | Restored layer contents

| Exit Code | Result|
|-----------|-------|
| `0`       | Success
| `11`      | Platform API incompatibility error
| `12`      | Buildpack API incompatibility error
| `1-10`, `13-99` | Generic lifecycle errors
| `300-399` | Restoration-specific lifecycle errors

> - **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer analysis.

Reviewer: ```suggestion
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.
```

> - **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer analysis.

Reviewer: ```suggestion
- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer analysis.
```
We also need to modify the data format for `io.buildpacks.lifecycle.metadata`[https://github.com/buildpacks/spec/blob/main/platform.md#iobuildpackslifecyclemetadata-json] to included persistent metadata in the format currently expected by the lifecycle.

Author: What has changed here? Sorry, I don't see what needs to be modified.

> - **Else** the lifecycle MUST analyze any app image layers or cached layers created by any buildpack present in the provided `<group>`.

Reviewer: ```suggestion
- **Else** the lifecycle MUST preform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.
```

Reviewer 2: preform -> perform

##### Layer Restoration

> When analyzing a given layer the lifecycle SHALL:

Reviewer: ```suggestion
When restoring a given layer the lifecycle SHALL:
```

| `<uid>`        | `CNB_USER_ID`         |                          | UID of the stack `User`
| `<skip-layers>`| `CNB_SKIP_LAYERS`     | `false`                  | Do not perform [layer restoration]((#layer-restoration)

##### Outputs
| Output                                | Description
|---------------------------------------|----------------------------------------------
| [exit status]                         | (see Exit Code table below for values)
| `/dev/stdout`                         | Logs (info)
| `/dev/stderr`                         | Logs (warnings, errors)
| `<layers>/<buidpack-id>/store.toml`   | Persistent metadata (see data format in [Buildpack Interface Specification](buildpack.md))
| `<layers>/<buidpack-id>/<layer>.toml` | Files containing the layer content metadata of each analyzed layer (see data format in [Buildpack Interface Specification](buildpack.md))
| `<layers>/<buidpack-id>/<layer>/*`.   | Restored layer contents

| Exit Code | Result|
|-----------|-------|
| `0`       | Success
| `11`      | Platform API incompatibility error
| `12`      | Buildpack API incompatibility error
| `1-10`, `13-99` | Generic lifecycle errors

> | `300-399` | Restoration-specific lifecycle errors

Reviewer 4: ```suggestion
| `40-49` | Restoration-specific lifecycle errors
```

Update error codes to match #196

- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.
- **Else** the lifecycle MUST perform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.

##### Layer Restoration
When restoring a given layer the lifecycle SHALL:
- **If** `build=true`, `cache=false`:
    - Do nothing
- **Else if** `launch=true`:
    - Write layer metadata read from the analyzed image to `<layers>/<buildpack-id>/<layer-name>.toml`

>     - Write the layer diffID from the analyzed image to `<layers>/<buildpack-id>/<layer-name>.sha`

Reviewer: ```suggestion
```
We don't need the `.sha` files anymore.

Author: Removed in 6480d2e975ba81d5b61d681f0177539408c834b8

- **Else if** `cache=true` AND the cache DOES NOT contain a layer with matching diffID
    - Must not write layer metadata
- **Else if** `cache=true`:

>     - Write layer metadata read from the cache to `<layers>/<buildpack-id>/<layer-name>.toml`

Reviewer: Seems like we are missing a description of when to restore cached layer contents? In the interest of normalized independent API maybe we should just refer out to https://github.com/buildpacks/spec/blob/main/buildpack.md#layer-types for a description of what to restore when. Then, any future changes could be a buildpack API concern. We merely need to say that the lifecycle must use the provided `cache-dir` or `cache-image` to retrieve layer contents.

Author: I took a stab at removing all the detail here, but I might've removed too much. Let me know what you think. Fixed in 6480d2e975ba81d5b61d681f0177539408c834b8

| [exit status]                         | (see Exit Code table below for values)
| `/dev/stdout`                         | Logs (info)
| `/dev/stderr`                         | Logs (warnings, errors)
| `<layers>/<buidpack-id>/store.toml`   | Persistent metadata (see data format in [Buildpack Interface Specification](buildpack.md))
| `<layers>/<buidpack-id>/<layer>.toml` | Files containing the layer content metadata of each analyzed layer (see data format in [Buildpack Interface Specification](buildpack.md))
| `<layers>/<buidpack-id>/<layer>/*`.   | Restored layer contents

| Exit Code | Result|
|-----------|-------|
| `0`       | Success
| `11`      | Platform API incompatibility error
| `12`      | Buildpack API incompatibility error
| `1-10`, `13-99` | Generic lifecycle errors
| `40-49` | Restoration-specific lifecycle errors

- For each buildpack in `<group>`, if persistent metadata for that buildpack exists in the analysis metadata, lifecycle MUST write a toml representation of the persistent metadata to `<layers>/<buildpack-id>/store.toml`
- **If** `<skip-layers>` is `true` the lifecycle MUST NOT perform layer restoration.

> - **Else** the lifecycle MUST perform [layer restoration](#layer-restoration) for any app image layers or cached layers created by any buildpack present in the provided `<group>`.

Reviewer 2: Does this apply to app image layers? `#layer-restoration` only talks about the cache.

    - Write layer metadata read from the cache to `<layers>/<buildpack-id>/<layer-name>.toml`

>     - Write the layer diffID from the cache to `<layers>/<buildpack-id>/<layer-name>.sha`

Reviewer: ```suggestion
```

For each layer metadata file found in the `<layers>` directory, the lifecycle:
- MUST restore cached layer contents if the cache contains a layer with matching diffID

> - MUST remove layer metadata if `cache=true` AND the cache DOES NOT contain a layer with matching diffID

Reviewer: Instead of require the `restorer` to write the layer metadata and then remove it. We should combine these lines with the lines above and say that the restorer must not write the layer memtadata in these cases.

#### `analyzed.toml` (TOML)

```toml
[previous-image]
  reference = "<image reference>"
[previous-image.metadata]
  # previous image layer metadata
[run-image]
  reference = "<image reference>"
  mixins = ["libgc", "libpq"]

> [build-image]
>   mixins = ["jq", "libgc", "libpq"]

Reviewer: ```suggestion
[build-image]
  stack-id = "<string>"
  mixins = ["jq", "libgc", "libpq"]
```
I want to add this meanly for symmetry with `stack.toml`. It could be useful eventually if we wanted to ensure that the `detector` and `builder` are in fact running on stack provided to `analyzer`.

Reviewer 3: Should we bubble this up so `stack-id` is shared between build and run? When I tried implementing this in pack, it was a little awkward because we don't usually scope a stack ID to only build or run.

Reviewer: @jkutner I could get behind that idea. We should still check that the run-image actually has the declared stack ID though.

Author: I added `stack-id` in 6480d2e975ba81d5b61d681f0177539408c834b8

Reviewer 3: Let's go with `[build-image] stack-id`, in this PR and I'll open a follow-up PR to change (then we can debate that separately from this PR).

[run-image]
 image = "<image>"
 mirrors = ["<mirror>", "<mirror>"]
[build-image]
 stack-id = "<string>"

>  mixins = ["jq", "libgc", "libpq"]

Reviewer: ```suggestion
 mixins = ["<mixin name>"]
```
`jq` etc. are very example-y and we generally don't do that in the data formats. See https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml

Author: Removed in 6480d2e975ba81d5b61d681f0177539408c834b8
